lib/Dialect/TTIR/Analysis/LegalGridAnalysis.cpp

```diff
@@ -3,13 +3,143 @@
+                   "row_major", // todo: expose parameter to LayoutAttr
+                   false))
+             : std::nullopt;
+}
+
+const std::string memref_get_buffer_type_str(const mlir::MemRefType &memref) {
+  return ::mlir::tt::MemorySpaceEnumToString(
+             mlir::cast<tt::MemorySpaceAttr>(memref.getMemorySpace())
+                 .getValue())
+      .str();
+}
+
+const std::string
+layout_attr_get_tensor_memory_layout_str(const mlir::tt::LayoutAttr &layout) {
+  return ::mlir::tt::TensorMemoryLayoutEnumToString(layout.getMemLayout())
+      .str();
+}
+
+ttnn::mlir_interface::memory_config_tuple
+layout_get_memory_config(const mlir::tt::LayoutAttr &layout) {
+  std::string tensor_memory_layout =
+      layout_attr_get_tensor_memory_layout_str(layout);
+  std::string buffer_type = memref_get_buffer_type_str(layout.getMemref());
+  auto shard_spec = layout_get_shard_spec(layout);
+  return std::make_tuple(tensor_memory_layout, buffer_type, shard_spec);
+}
+
+std::string memref_get_element_type(const mlir::MemRefType &memref) {
+  mlir::Type element_type = memref.getElementType();
+  // what's better way to to this?
+  // auto data_type = mlir::cast<DataType>(element_type);
+  std::string ret_value;
+  llvm::raw_string_ostream os(ret_value);
+  element_type.print(os);
+  return os.str();
+}
+} // namespace mlir::ttnn_wrapper
+
 namespace mlir::tt::ttir {
 
 bool mock_is_output_tensor_legal_for_op(Operation *op, LayoutAttr layout) {
   // Placeholder, needs to be replaced with a call the the TTNN op interface.
-  return true;
+
+  // serialize mlir structures to interface structures
+  auto memref = layout.getMemref();
+  auto shape = ttnn_wrapper::memref_get_tensor_shape(memref);
+  std::string data_type = ttnn_wrapper::memref_get_element_type(memref);
+  ttnn::mlir_interface::memory_config_tuple memory_config =
+      ttnn_wrapper::layout_get_memory_config(layout);
+
+  // call mlir_interface library per op name
+  auto op_name_str = op->getName().getStringRef().str();
+  bool is_valid = false;
+  if (op_name_str == "ttir.multiply" || op_name_str == "ttir.add" ||
+      op_name_str == "ttir.subtract") {
+    is_valid =
+        ttnn::mlir_interface::does_binary_op_support_input_output_constraints(
+            shape, memory_config, data_type, shape, memory_config, data_type,
+            memory_config, data_type);
+  } else if (op_name_str == "ttir.softmax") {
+    is_valid =
+        ttnn::mlir_interface::does_softmax_op_support_input_output_constraints(
+            shape, memory_config, data_type, shape, memory_config, data_type);
+  } else if (op_name_str == "ttir.relu") {
+    is_valid =
+        ttnn::mlir_interface::does_unary_op_support_input_output_constraints(
+            "RELU", // todo agree upon mapping to ttnn::mlir_interface
+            shape, memory_config, data_type, shape, memory_config, data_type);
+  } else {
+    llvm::outs() << op->getName() << " missing ttnn interface\n";
+    return false;
+  }
```

Comment:
Is there a nicer way of doing op class dedution - in TTIR there are base classes for each op category (e.g., `TTIR_ElementwiseBinaryOp`, `TTIR_ElementwiseUnaryOp`, ...), can we use that info to get the op class?

Reply:
It is possible to do `llvm::isa<MatmulOp>(op)`.

Reply:
Yes this is the way, thanks!